### PR TITLE
V1.2.0

### PR DESCRIPTION
**Pull request by daviesdoclc**
Implement refresh tokens

### CODE REVIEW

File: www/OCLCAuthenticationPlugin.js [hunk 1]
```diff
@@ -470,24 +453,86 @@ function _authorize(options, success, failure) {
...
+        };
         xhr.onerror = function() {
             failure();
-        }
+        };
+        xhr.send();
+    }
+}
+
+function _refresh(options, success, failure) {
```
Comment (daviesdoclc) [inline www/OCLCAuthenticationPlugin.js]: really just concentrate on this function.  the rest was cleanup and restructuring.

**Comment by oclc-platform-dev** [comment]
I am able to login and see this pull request.

File: www/OCLCAuthenticationPlugin.js [hunk 2]
```diff
@@ -423,32 +404,34 @@ function _readTextFileAsBase64(url, callback) {
...
 
+    if (!options || !success || !failure) {
+        console.log("AUTHENTICATION ABORTED - OCLC Authentication Plugin called without required paramters");
+        return;
+    }
+
     // if no locale specified then default to english
     if (!options.locale) {
         options.locale = "en-US";
     }
```
Comment (campbelg) [inline www/OCLCAuthenticationPlugin.js]: This seems like a dangerous change. If options is null, the javascript crashes.

File: www/OCLCAuthenticationPlugin.js [hunk 3]
```diff
@@ -423,32 +404,34 @@ function _readTextFileAsBase64(url, callback) {
...
             "token_type": "bearer",
-            "expires_in": "9999", "expires_at": "3000-01-01 00:00:00Z"
+            "expires_in": "1200",
+            "expires_at": (new Date()).getTime() + (1200 * 1000) - 60000,
+            "refresh_token": "rt_SomeRefreshToken",
+            "refresh_token_expires_in": "0",
+            "refresh_token_expires_at": "0"
         });
 
     } else {
```
Comment (campbelg) [inline www/OCLCAuthenticationPlugin.js]: See here we handle options not existing.

File: www/OCLCAuthenticationPlugin.js [hunk 4]
```diff
@@ -470,24 +453,87 @@ function _authorize(options, success, failure) {
...
+    }
+}
+
+function _refresh(options, success, failure) {
+
+    if (!options || !success || !failure) {
+        console.log("AUTHENTICATION ABORTED - OCLC Authentication Plugin called without required paramters");
+        return;
+    }
+
```
Comment (campbelg) [inline www/OCLCAuthenticationPlugin.js]: options || options.mock in case options === null or typeof options === "undefined"

**Pull request merged**